[Merged by Bors] - feat(algebra/opposites): Provide semimodule instances and op_linear_equiv

eric-wieser opened the pull request:
We already have a `has_scalar` definition via an `algebra` definition.

The definition used there satisfies a handful of other typeclasses too, and also allows for `op_add_equiv` to be stated more strongly as `op_linear_equiv`.

This also cuts back the imports a little on `algebra.module.basic`, which means formerly-transitive imports have to be added to downstream files.

---
<!--
put comments you want to keep out of the PR commit here.
If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
-->

- [x] depends on: #4936, which adjusts the import order to avoid loops here

This cleans up some imports, which will lead to CI failures until I track down all the missing transitive imports downstream

eric-wieser commented on src/algebra/module/opposites.lean:
```diff
@@ -0,0 +1,47 @@
+  ..opposite.distrib_mul_action M R }
+
+/-- The function `op` is a linear equivalence. -/
+@[simps apply]
```
```suggestion
```

Vierkantor commented on src/algebra/algebra/basic.lean:
```diff
@@ -327,12 +327,11 @@ namespace opposite
-@[simp] lemma op_smul (c : R) (a : A) : op (c • a) = c • op a := rfl
+{ to_ring_hom := (algebra_map R A).to_opposite $ λ x y, algebra.commutes _ _,
+  smul_def' := λ c x, unop_injective $ by {
+    dsimp, simp only [op_mul, algebra.smul_def, algebra.commutes, op_unop]},
```
```suggestion
  smul_def' := λ c x, unop_injective $
    by { dsimp, simp only [op_mul, algebra.smul_def, algebra.commutes, op_unop] },
```

eric-wieser replied:
What's the heuristic you're using here? I'm never quite sure how to wrap these things.

Vierkantor commented on src/algebra/algebra/basic.lean:
```diff
@@ -327,12 +327,11 @@ namespace opposite
+  smul_def' := λ c x, unop_injective $ by {
+    dsimp, simp only [op_mul, algebra.smul_def, algebra.commutes, op_unop]},
+  commutes' := λ r, op_induction $ λ x, by dsimp; simp only [← op_mul, algebra.commutes],
+  ..(infer_instance : has_scalar R Aᵒᵖ) }
```
I suspect that hiding the instance behind `infer_instance` gives problems with unification later on.
```suggestion
  ..opposite.has_scalar A R }
```

Vierkantor reviewed (approved):
Looks good to me.

bors d+

eric-wieser committed:
feat(algebra/opposites): Provide semimodule instances and op_linear_equiv (#4954)

We already have a `has_scalar` definition via an `algebra` definition.

The definition used there satisfies a handful of other typeclasses too, and also allows for `op_add_equiv` to be stated more strongly as `op_linear_equiv`.

This also cuts back the imports a little on `algebra.module.basic`, which means formerly-transitive imports have to be added to downstream files.

eric-wieser committed:
Add missing import
